server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java

```diff
@@ -0,0 +1,286 @@
+        assertThat(mailQueueItem.getMail().getLastUpdated())
+            .isEqualTo(lastUpdated);
+    }
+
+    @Test
+    default void queueShouldPreserveName() throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name")
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem = getMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getName())
+            .isEqualTo("name");
+    }
+
+    @Test
+    default void queueShouldPreservePerRecipientHeaders() throws Exception {
+        PerRecipientHeaders.Header header = PerRecipientHeaders.Header.builder()
+            .name("any")
+            .value("any")
+            .build();
+        getMailQueue().enQueue(defaultMail()
+            .addHeaderForRecipient(header, RECIPIENT1)
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem = getMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getPerRecipientSpecificHeaders()
+            .getHeadersForRecipient(RECIPIENT1))
+            .containsOnly(header);
+    }
+
+    @Test
+    default void dequeueShouldBeFifo() throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name1")
+            .build());
+        getMailQueue().enQueue(defaultMail()
+            .name("name2")
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        mailQueueItem1.done(true);
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem2.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo("name1");
+        assertThat(mailQueueItem2.getMail().getName()).isEqualTo("name2");
+    }
+
+    @Test
+    default void dequeueCouldBeInterleaving() throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name1")
+            .build());
+        getMailQueue().enQueue(defaultMail()
+            .name("name2")
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem1.done(true);
+        mailQueueItem2.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo("name1");
+        assertThat(mailQueueItem2.getMail().getName()).isEqualTo("name2");
+    }
+
+
+    @Test
+    default void dequeueCouldBeInterleavingWithOutOfOrderAck() throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name1")
+            .build());
+        getMailQueue().enQueue(defaultMail()
+            .name("name2")
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem2.done(true);
+        mailQueueItem1.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo("name1");
```

Comment:
I don't understand why you use this assertion to check the state of your queue. What does it tell you about the result of `done`calls ?

Reply:
Sorry I just ensure the dequeued mail identity... What would you expect?

Reply:
Ok so, let's imagine the queue doesn't support calling done in the wrong order and decide to re-enqueue mail1 when you call mail2.done(true). Your assertion doesn't checks that.
You should check that the queue is empty, for example.